File: src/features/intelligence/services/DemandEngine.js
/**
 * DEMAND ENGINE
 * 
 * A deterministic business intelligence service.
 * Synthesizes raw RetailerProfiles into geographical product and category DemandSignals.
 */

const PRODUCT_DICTIONARY = {
  'Dairy': ['paneer', 'milk', 'butter', 'curd', 'ghee'],
  'Staples': ['atta', 'rice', 'dal', 'sugar', 'salt'],
  'Beverages': ['water', 'juice', 'soft drink', 'cold drink'],
  'Snacks': ['biscuits', 'namkeen', 'chips', 'snacks'],
  'Spices': ['masala', 'turmeric', 'chilli', 'cumin'],
  'Household': ['detergent', 'soap', 'cleaner']
};

const SCORING_WEIGHTS = {
  EXPLICIT_UNMET_NEED: 3,
  PRODUCT_REQUIREMENT: 2,
  REQUESTED_CATEGORY: 1
};

export const DemandEngine = {
  
  /**
   * Main analysis function.
   * Processes an array of retailer profiles and returns normalized demand signals.
   * 
   * @param {Array} retailers - Array of RetailerProfile objects.
   * @param {Object} options - Reserved for future configurations.
   * @returns {Object} { categorySignals, productSignals }
   */
  analyze: (retailers = [], options = {}) => {
    if (!Array.isArray(retailers)) return { categorySignals: [], productSignals: [] };

    const categoryMap = new Map(); // Key: 'location|Category'
    const productMap = new Map();  // Key: 'location|Product'

    // Helper to extract location grouping key
    const getLocationKey = (location) => {
      if (!location) return 'unknown';
      const area = location.area?.trim();
      const district = location.district?.trim();
      
      if (area && district) return `${area}, ${district}`;
      if (area) return area;
      if (district) return district;
      return 'unknown';
    };

    // Helper to mutate map
    const addSignal = (map, locationKey, itemType, itemName, retailerId, weight, source, parentCategory = null) => {
      const key = `${locationKey}|${itemName}`;
      
      if (!map.has(key)) {
        map.set(key, {
          locationKey,
          type: itemType, // 'category' or 'product'
          name: itemName,
          category: parentCategory || (itemType === 'category' ? itemName : undefined),
          retailerSet: new Set(),
          signalStrength: 0,
          sourceBreakdown: {
            requestedCategory: 0,
            unmetNeed: 0,
            requirementMention: 0
          }
        });
      }

      const entry = map.get(key);
      
      // Accumulate total signal strength from all mentions
      entry.signalStrength += weight;
      entry.sourceBreakdown[source] += 1;
      
      // Deduplication: Add to set (ensures 1 retailer counts as exactly 1 "retailerCount")
      if (retailerId) {
        entry.retailerSet.add(retailerId);
      }
    };

    // Process each retailer
    retailers.forEach(retailer => {
      // Tolerate missing or malformed data
      const id = retailer.id;
      const locationKey = getLocationKey(retailer.location);
      const demandedCats = Array.isArray(retailer.demandedCategories) ? retailer.demandedCategories : [];
      
      // Handle unmetNeeds format safely
      let unmetCategories = [];
      let unmetOther = '';
      if (typeof retailer.unmetNeeds === 'object' && retailer.unmetNeeds !== null) {
        unmetCategories = Array.isArray(retailer.unmetNeeds.categories) ? retailer.unmetNeeds.categories : [];
        unmetOther = retailer.unmetNeeds.other || '';
      } else if (typeof retailer.unmetNeeds === 'string') {
        unmetOther = retailer.unmetNeeds;
      }
      
      const requirementsText = (retailer.requirements || '').toLowerCase();
      const unmetOtherText = unmetOther.toLowerCase();

      // 1. Process Demanded Categories (Weight 1)
      demandedCats.forEach(cat => {
        addSignal(categoryMap, locationKey, 'category', cat, id, SCORING_WEIGHTS.REQUESTED_CATEGORY, 'requestedCategory');
      });

      // 2. Process Unmet Need Categories (Weight 3)
      unmetCategories.forEach(cat => {
        addSignal(categoryMap, locationKey, 'category', cat, id, SCORING_WEIGHTS.EXPLICIT_UNMET_NEED, 'unmetNeed');
      });

      // 3. Extract Explicit Products from Text Requirements & Unmet 'Other' (Weight 2 for requirements, 3 for unmet)
      Object.entries(PRODUCT_DICTIONARY).forEach(([category, keywords]) => {
        keywords.forEach(keyword => {
          // Check in requirements
          if (requirementsText.includes(keyword)) {
            // Capitalize product keyword
            const productName = keyword.charAt(0).toUpperCase() + keyword.slice(1);
            addSignal(productMap, locationKey, 'product', productName, id, SCORING_WEIGHTS.PRODUCT_REQUIREMENT, 'requirementMention', category);
            // Implicitly add to category
            addSignal(categoryMap, locationKey, 'category', category, id, SCORING_WEIGHTS.PRODUCT_REQUIREMENT, 'requirementMention');
          }
          // Check in unmet needs text
          if (unmetOtherText.includes(keyword)) {
            const productName = keyword.charAt(0).toUpperCase() + keyword.slice(1);
            addSignal(productMap, locationKey, 'product', productName, id, SCORING_WEIGHTS.EXPLICIT_UNMET_NEED, 'unmetNeed', category);
            addSignal(categoryMap, locationKey, 'category', category, id, SCORING_WEIGHTS.EXPLICIT_UNMET_NEED, 'unmetNeed');
          }
        });
      });
    });

    // Helper to determine demand level based on thresholds
    const getDemandLevel = (signalStrength, retailerCount) => {
      if (signalStrength >= 10 || retailerCount >= 5) return 'High';
      if (signalStrength >= 4 || retailerCount >= 2) return 'Medium';
      return 'Low';
    };

    // Helper to finalize mapped signals
    const finalizeSignals = (map) => {
      return Array.from(map.values()).map(entry => {
        const retailerCount = entry.retailerSet.size;
        
        let area = '';
        let district = '';
        
        if (entry.locationKey !== 'unknown') {
          const parts = entry.locationKey.split(', ');
          if (parts.length === 2) {
            area = parts[0];
            district = parts[1];
          } else {
            // In a real app we'd need better tracking of what the single string is
            district = parts[0]; 
          }
        }

        return {
          id: `demand_${entry.name.toLowerCase().replace(/ /g, '_')}_${entry.locationKey.toLowerCase().replace(/[^a-z0-9]/g, '_')}`,
          area: area || entry.locationKey,
          district: district,
          [entry.type]: entry.name,
          category: entry.category,
          retailerCount,
          demandLevel: getDemandLevel(entry.signalStrength, retailerCount),
          signalStrength: entry.signalStrength,
          sourceBreakdown: entry.sourceBreakdown
        };
      }).sort((a, b) => b.signalStrength - a.signalStrength); // Highest demand first
    };

    return {
      categorySignals: finalizeSignals(categoryMap),
      productSignals: finalizeSignals(productMap)
    };
  }
};

File: backend/app/modules/intelligence/services/demand_engine.py
from sqlalchemy.orm import Session
from sqlalchemy.dialects.postgresql import insert
from app.models.profiles import RetailerProfile
from app.models.intelligence import DemandSignal
from app.modules.intelligence.services.normalization import NormalizationService

import uuid

class DemandEngine:
    PRODUCT_DICTIONARY = {
        'Dairy': ['paneer', 'milk', 'butter', 'curd', 'ghee'],
        'Staples': ['atta', 'rice', 'dal', 'sugar', 'salt'],
        'Beverages': ['water', 'juice', 'soft drink', 'cold drink'],
        'Snacks': ['biscuits', 'namkeen', 'chips', 'snacks'],
        'Spices': ['masala', 'turmeric', 'chilli', 'cumin'],
        'Household': ['detergent', 'soap', 'cleaner']
    }

    SCORING_WEIGHTS = {
        'EXPLICIT_UNMET_NEED': 3.0,
        'PRODUCT_REQUIREMENT': 2.0,
        'REQUESTED_CATEGORY': 1.0
    }
    
    @staticmethod
    def generate_id(retailer_id: str, product_id: str, category_id: str, signal_type: str, source: str) -> str:
        """Generates a deterministic ID for a DemandSignal based on its canonical uniqueness."""
        id_str = f"{retailer_id or ''}:{product_id or ''}:{category_id or ''}:{signal_type}:{source}"
        return str(uuid.uuid5(uuid.NAMESPACE_OID, id_str))

    @staticmethod
    def generate_demand(db: Session):
        """
        Reads retailer profiles, extracts demand, and idempotently upserts to DemandSignals.
        """
        retailers = db.query(RetailerProfile).all()
        signals = []

        for retailer in retailers:
            location_id = retailer.location_id
            retailer_id = retailer.id

            # 1. Demanded Categories
            demanded_cats = retailer.demanded_categories or []
            for cat in demanded_cats:
                cat_id = NormalizationService.map_category_id(db, cat)
                signals.append({
                    "id": DemandEngine.generate_id(retailer_id, None, cat_id, "category_requirement", cat),
                    "retailer_id": retailer_id,
                    "product_id": None,
                    "category_id": cat_id,
                    "location_id": location_id,
                    "signal_type": "category_requirement",
                    "source": cat,
                    "confidence": DemandEngine.SCORING_WEIGHTS['REQUESTED_CATEGORY']
                })

            # 2. Unmet Needs Categories
            unmet_needs = retailer.unmet_needs or {}
            if isinstance(unmet_needs, dict):
                unmet_cats = unmet_needs.get('categories', [])
                unmet_other = unmet_needs.get('other', '')
            else:
                unmet_cats = []
                unmet_other = str(unmet_needs) if unmet_needs else ""

            for cat in unmet_cats:
                cat_id = NormalizationService.map_category_id(db, cat)
                signals.append({
                    "id": DemandEngine.generate_id(retailer_id, None, cat_id, "unmet_need", cat),
                    "retailer_id": retailer_id,
                    "product_id": None,
                    "category_id": cat_id,
                    "location_id": location_id,
                    "signal_type": "unmet_need",
                    "source": cat,
                    "confidence": DemandEngine.SCORING_WEIGHTS['EXPLICIT_UNMET_NEED']
                })

            # 3. Requirements (Text/Array parsing)
            reqs = retailer.requirements or []
            if isinstance(reqs, list):
                reqs_text = " ".join(reqs).lower()
            else:
                reqs_text = str(reqs).lower()

            unmet_other_text = unmet_other.lower()

            for category, keywords in DemandEngine.PRODUCT_DICTIONARY.items():
                for keyword in keywords:
                    cat_id = NormalizationService.map_category_id(db, category)
                    prod_id = NormalizationService.map_product_id(db, keyword)

                    if keyword in reqs_text:
                        signals.append({
                            "id": DemandEngine.generate_id(retailer_id, prod_id, cat_id, "product_requirement", keyword),
                            "retailer_id": retailer_id,
                            "product_id": prod_id,
                            "category_id": cat_id,
                            "location_id": location_id,
                            "signal_type": "product_requirement",
                            "source": keyword,
                            "confidence": DemandEngine.SCORING_WEIGHTS['PRODUCT_REQUIREMENT']
                        })
                    
                    if keyword in unmet_other_text:
                        signals.append({
                            "id": DemandEngine.generate_id(retailer_id, prod_id, cat_id, "unmet_need", keyword),
                            "retailer_id": retailer_id,
                            "product_id": prod_id,
                            "category_id": cat_id,
                            "location_id": location_id,
                            "signal_type": "unmet_need",
                            "source": keyword,
                            "confidence": DemandEngine.SCORING_WEIGHTS['EXPLICIT_UNMET_NEED']
                        })

        # Idempotent insert via PostgreSQL ON CONFLICT
        # Ensure we have data
        if not signals:
            return {"status": "success", "inserted": 0, "processed": len(retailers)}

        # Need to deduplicate within the batch itself to avoid constraint errors
        unique_signals = {}
        for s in signals:
            key = s['id']
            # Prefer higher confidence if there are duplicates for some reason
            if key not in unique_signals or s['confidence'] > unique_signals[key]['confidence']:
                unique_signals[key] = s
        
        insert_stmt = insert(DemandSignal).values(list(unique_signals.values()))
        
        # On conflict do nothing for identical signals
        on_conflict_stmt = insert_stmt.on_conflict_do_nothing(
            index_elements=['id']
        )
        
        db.execute(on_conflict_stmt)
        db.commit()

        return {"status": "success", "processed_retailers": len(retailers), "signals_generated": len(unique_signals)}
